tools/codediff/run_command.sh

```diff
@@ -0,0 +1,112 @@
+#!/bin/bash
+# SPDX-FileCopyrightText: Copyright (c) 2023-present NVIDIA CORPORATION & AFFILIATES.
+# All rights reserved.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+# run_command.sh - run a single command and record environment info
+#
+set -e
+set -o pipefail
+usage() {
+echo "Usage: $0 [-h] [-n <run_name>] -o <output_directory> -- command to run and arguments"
+}
+while getopts "n:o:h" arg
+do
+case $arg in
+n)
+runname=$OPTARG
+;;
+o)
+testdir=$OPTARG
+;;
+h | *)
+usage
+exit 1
+;;
+esac
+done
+if [[ -z $testdir ]]
+then
+echo "The output directory must be specified with -o"
+usage
+exit 1
+fi
+# getopts stops parsing if it sees "--" but does not shift. We can detect that case and record command
+while [[ $# -gt 0 ]]
+do
+if [[ "$1" == "--" ]]
+then
+shift
+break
+fi
+shift
+done
+testcmd=$*
+if [[ -z "$testcmd" ]]
+then
+usage
+exit 1
+fi
+if [[ -f "$testdir/command" ]]
+then
+echo -n "Skipping since $testdir/command exists. "
+echo "To re-run, remove $testdir or specify another and try again."
+exit 1
+fi
+mkdir -p "$testdir"
+movecudafiles() {
+mkdir -p "$1/cuda" "$1/ptx"
+find . -maxdepth 1 -name '__tmp_kernel*.cu' -print0 | xargs -0 mv -t "$1/cuda"
+find . -maxdepth 1 -name '__tmp_kernel*.ptx' -print0 | xargs -0 mv -t "$1/ptx"
+}
+removecudafiles() {
+tmpdir="./.nvfuser_run_command_tmp"
+mkdir -p "$tmpdir"
+movecudafiles "$tmpdir"
+rm -rf "$tmpdir"
+}
+stdoutfile="$testdir/incomplete-stdout-$(date +%Y%m%d_%H%M%S).log"
+cleanup() {
+numcu=$(find . -maxdepth 1 -name '__tmp_kernel*.cu' | wc -l)
+numptx=$(find . -maxdepth 1 -name '__tmp_kernel*.ptx' | wc -l)
+if (( numcu + numptx > 0 ))
+then
+echo "Interrupted. Removing $numcu temporary .cu files and $numptx temporary .ptx files"
+removecudafiles
+fi
+logbase=$(basename "$stdoutfile")
+# strip incomplete- from base name
+completelogbase=${logbase#incomplete-}
+mv "$stdoutfile" "$testdir/$completelogbase" 2> /dev/null || true
+}
+trap "cleanup" EXIT
+# Allow command to fail, but record exit code
+set +e
+$testcmd | tee "$stdoutfile"
+echo $? > "$testdir/exitcode"
```

Comment:
https://unix.stackexchange.com/questions/14270/get-exit-status-of-process-thats-piped-to-another

For the exit code of a bash pipe, we may want to use `${PIPESTATUS[0]}` for the first command

Reply:
Ah right. Good catch!